functorch/experimental/_scan.py

```diff
@@ -0,0 +1,264 @@
+        elif all(leaf is None for leaf in leaves):
+            # Backward graph can return None output when forward inputs doesn't require grad.
+            # When we eagerly execute backward graph, we need to call _stack_pytree on its output,
+            # therefore we need to deal with None output.
+            stacked_out.append(None)
+        else:
+            raise RuntimeError(f"Cannot stack {leaves}.")
+    return pytree.tree_unflatten(stacked_out, out_spec)
+
+@scan_impl.py_impl(DispatchKey.CompositeExplicitAutograd)
+def scan_dense(f, flat_init, flat_xs, ys_grad=None, carries=None, return_all_carries=False, reverse=False):
+    '''
+        The scan_dense implementation is reused for executing both forward and backward graph.
+        When executing forward graph, f is the fwd_graph, and flat_init, flat_xs are the typical flattened
+        init and xs lists, and ys_grad, carries should be None, reverse should be False (not exposing reverse argument at the top level yet).
+        Note that the return_all_carries argument will control whether we return all the intermediate carries as the third
+        output (except the final output carry, which is returned as the first output)
+
+        When executing backward graph, f is the bwd_graph, and flat_init should be the initial output carry gradient, flat_xs is still
+        the same flattened input xs. ys_grad is the output gradient, and carries is a list of all the intermediate carries generated during
+        forward pass (less the final output carry). This list of carries are needed for computing the backward gradient at each iteration.
+        When running bwd_graph with scan_dense, we should set reverse = True because the gradients are computed bottom up
+    '''
+    carry = flat_init
+    # each carry output is needed for calculating gradients backwards
+    out_carries = []
+    out_pytrees = []
+    direction = -1 if reverse else 1
+    xs_unstacked = _unstack_pytree(flat_xs)[::direction]
+    iterable_input = zip(xs_unstacked)
+    if carries is not None and ys_grad is not None:
+        # suppose both are not none, that means we are reusing the scan_dense for gradient computation in bwd pass
+        # in this case, the flat_init is actually the gradients on final carry output from fwd pass
+        ys_grad_unstacked = _unstack_pytree(ys_grad)[::direction]
+        iterable_input = zip(xs_unstacked, carries[::direction], ys_grad_unstacked)
+    for inp in iterable_input:
+        # saves the initial carry input for each iteration
+        out_carries.append(carry)
+        carry, flattened_out = f(carry, *inp)
+        out_pytrees.append(flattened_out)
+    ys = _stack_pytree(out_pytrees[::direction])
+    if return_all_carries:
+        return carry, ys, out_carries
+    return carry, ys
+
+@scan_impl.py_impl(DispatchKey.Autograd)
+def scan_autograd(f, flat_init, flat_xs, ys_grad=None, carries=None, return_all_carries=False, reverse=False):
```

Comment:
Why do we need ys_grad and carries keyword argument? Can we compact them as part of flat_xs?